### PR TITLE
Fix unbounded heap growth in the priority mempool.

### LINKED ISSUE #8775: Persistent memory growth when using the priority mempool
When running a testnet against Tendermint v0.35, the Vega team observed that enabling the new priority mempool feature triggered unexpected memory growth in their nodes as compared to using the legacy (sequential) mempool on a similar workload. In summary, a validator with a 10K mempool with a 2K cache shows moderate growth on a resident heap of between 500MiB and 1GiB using the sequential mempool, but quickly rose to 3GiB (which ran over cap and terminated the process) when using the priority mempool.

### PR DESCRIPTION
**Pull request by creachadair**
### Description

The primary effect of this change is to simplify the implementation of the
priority mempool to eliminate an unbounded heap growth observed by Vega team
when it was enabled in their testnet. It updates and fixes #8775.

The main body of this change is to remove the auxiliary indexing structures,
and use only the concurrent list structure (the same as the legacy mempool) to
maintain both gossip order and priority.

This means that operations that require priority information, such as block
updates and insert-time evictions, require a linear scan over the mempool.
This tradeoff greatly simplifies the code and eliminates the long-term heap
load, at the cost of some extra CPU and short-lived working memory during
CheckTx and Update calls.

On the existing CheckTx benchmarks, this cost is only about 3 microseconds
slower than the legacy mempool, and given that CheckTx rate is generally
limited by network latencies (order milliseconds), the overhead is negligible.

In addition:

- Rework synchronization so that access to shared data structures is safe.
  Previously shared locks were used to exclude block updates during calls that
  update mempool state. Now access is properly exclusive where necessary.

- Fix a bug in the recheck flow, where priority updates from the application
  were not correctly reflected in the index structures.

- Eliminate the need for separate recheck cursors during block update. This
  avoids the need to explicitly invalidate elements of the concurrent list,
  which averts the dependency cycle that led to objects being pinned.

- Clean up, clarify, and fix inaccuracies in documentation comments throughout
  the package.

### Note to Reviewers

This change passes all the existing tests in the package (other than those eliminated by removal of the data structures they exercised). However, the existing test coverage is relatively weak, and should be improved before this code is put into production.

### CODE REVIEW

**Comment by creachadair** [comment]
I think one of these test failures is real. I'll take a closer look in the morning.

File: internal/mempool/v1/mempool.go [hunk 1]
```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
...
 			txmp.logger.Error(
-				"rejected incoming good transaction; tx already exists for sender",
-				"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
+				"rejected valid incoming transaction; tx already exists for sender",
+				"tx", fmt.Sprintf("%X", w.tx.Hash()),
 				"sender", sender,
 			)
-			txmp.metrics.RejectedTxs.Add(1)
 			return
 		}
```
Comment (cmwaters) [inline internal/mempool/v1/mempool.go:518]: Can we append this error to `checkTxRes.CheckTx.MempoolError` so RPC callers understand that their tx failed and why

File: internal/mempool/v1/mempool.go [hunk 2]
```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
...
-	if len(sender) > 0 {
-		if wtx := txmp.txStore.GetTxBySender(sender); wtx != nil {
+	// Disallow multiple concurrent transactions from the same sender assigned
+	// by the ABCI application. As a special case, an empty sender is not
+	// restricted.
+	if sender != "" {
+		elt, ok := txmp.txBySender[sender]
+		if ok {
+			w := elt.Value.(*WrappedTx)
 			txmp.logger.Error(
```
Comment (cmwaters) [inline internal/mempool/v1/mempool.go:512]: Should this be an error? Doesn't seem like something a node operator should particularly care about

File: internal/mempool/v1/mempool.go [hunk 3]
```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
...
-				"rejected incoming good transaction; tx already exists for sender",
-				"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
+				"rejected valid incoming transaction; tx already exists for sender",
+				"tx", fmt.Sprintf("%X", w.tx.Hash()),
 				"sender", sender,
 			)
-			txmp.metrics.RejectedTxs.Add(1)
```
Comment (cmwaters) [inline internal/mempool/v1/mempool.go:555]: Why does this not count as a rejected tx?

File: internal/mempool/v1/mempool.go [hunk 4]
```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
...
 			txmp.cache.Remove(wtx.tx)
 			txmp.logger.Error(
-				"rejected incoming good transaction; mempool full",
+				"rejected valid incoming transaction; mempool is full",
 				"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
 				"err", err.Error(),
 			)
 			txmp.metrics.RejectedTxs.Add(1)
 			return
 		}
```
Comment (cmwaters) [inline internal/mempool/v1/mempool.go:549]: Can we add an error to `mempool_error` indicating that this transaction was dropped because the mempool is full and it is the lowest priority

File: internal/mempool/v1/mempool.go [hunk 5]
```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
...
+// Otherwise, if the mempool is full, check for lower-priority transactions
+// that can be evicted to make room for the new one. If no such transactions
+// exist, this transaction is logged and dropped; otherwise the selected
+// transactions are evicted.
 //
-// NOTE:
-// - An explicit lock is NOT required.
-func (txmp *TxMempool) initTxCallback(wtx *WrappedTx, res *abci.Response, txInfo mempool.TxInfo) {
+// Finally, the new transaction is added and size stats updated.
+func (txmp *TxMempool) initTxCallback(wtx *WrappedTx, res *abci.Response) {
```
Comment (cmwaters) [inline internal/mempool/v1/mempool.go:462]:
```suggestion
func (txmp *TxMempool) initialTxCallback(wtx *WrappedTx, res *abci.Response) {
```
nit: when I see init I'm too used to thinking initialize

File: internal/mempool/v1/mempool.go [hunk 6]
```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
...
-		if !txmp.txStore.IsTxRemoved(wtx.hash) {
-			_, err := txmp.proxyAppConn.CheckTxAsync(ctx, abci.RequestCheckTx{
-				Tx:   wtx.tx,
-				Type: abci.CheckTxType_Recheck,
-			})
-			if err != nil {
-				// no need in retrying since the tx will be rechecked after the next block
-				txmp.logger.Error("failed to execute CheckTx during rechecking", "err", err)
-			}
+		_, err := txmp.proxyAppConn.CheckTxAsync(ctx, abci.RequestCheckTx{
```
Comment (cmwaters) [inline internal/mempool/v1/mempool.go:708]: Can you add a comment here perhaps that this function will call `recheckTxCallback` which is responsible for removing the transactions else for the untrained reader it might not be clear that that is happening

File: internal/mempool/v1/mempool.go [hunk 7]
```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
...
-			// we reached the end of the recheckTx list without finding a tx
-			// matching the one we received from the ABCI application.
-			// Return without processing any tx.
-			txmp.recheckCursor = nil
+	numLeft := atomic.AddInt64(&txmp.txRecheck, -1)
+	if numLeft <= 0 {
+		defer txmp.notifyTxsAvailable()
+		if numLeft < 0 {
 			return
 		}
```
Comment (cmwaters) [inline internal/mempool/v1/mempool.go:641]: What does this do? I'm slightly confused

File: internal/mempool/v1/mempool.go [hunk 8]
```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
...
 	}
 
-	if txmp.recheckCursor == nil {
-		txmp.logger.Debug("finished rechecking transactions")
-
-		if txmp.Size() > 0 {
-			txmp.notifyTxsAvailable()
-		}
+	if checkTxRes.CheckTx.Code == abci.CodeTypeOK && err == nil {
+		wtx.SetPriority(checkTxRes.CheckTx.Priority)
```
Comment (cmwaters) [inline internal/mempool/v1/mempool.go:663]: I'm guessing it's not possible to change sender? If that's the case then maybe make a note of it

File: internal/mempool/v1/mempool.go [hunk 9]
```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
...
+		return // N.B. Size of mempool did not change
 	}
 
+	txmp.logger.Debug(
+		"existing transaction no longer valid; failed re-CheckTx callback",
+		"priority", wtx.Priority(),
+		"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
+		"err", err,
+		"code", checkTxRes.CheckTx.Code,
+	)
```
Comment (cmwaters) [inline internal/mempool/v1/mempool.go:673]: Do we want to report other information about the response i.e. codespace or log?

File: internal/mempool/v1/mempool.go [hunk 10]
```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
...
+// purgeExpiredTxs removes all transactions from the mempool that have exceeded
+// their respective height or time-based limits as of the given blockHeight.
+// Transactions removed by this operation are not removed from the cache.
 //
-// NOTE: purgeExpiredTxs must only be called during TxMempool#Update in which
-// the caller has a write-lock on the mempool and so we can safely iterate over
-// the height and time based indexes.
+// The caller must hold txmp.mtx exclusively.
 func (txmp *TxMempool) purgeExpiredTxs(blockHeight int64) {
 	now := time.Now()
```
Comment (cmwaters) [inline internal/mempool/v1/mempool.go:749]: nit: It may make it easier to test if time is an argument as well as blockHeight but YMMV

File: internal/mempool/v1/mempool.go [hunk 11]
```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
...
-// transaction will be removed from the mempool entirely, except for the cache.
+// purgeExpiredTxs removes all transactions from the mempool that have exceeded
+// their respective height or time-based limits as of the given blockHeight.
+// Transactions removed by this operation are not removed from the cache.
 //
-// NOTE: purgeExpiredTxs must only be called during TxMempool#Update in which
-// the caller has a write-lock on the mempool and so we can safely iterate over
-// the height and time based indexes.
+// The caller must hold txmp.mtx exclusively.
 func (txmp *TxMempool) purgeExpiredTxs(blockHeight int64) {
```
Comment (cmwaters) [inline internal/mempool/v1/mempool.go:748]: Maybe a premature optimisation but if we add 
```
if txmp.config.TTLNumBlocks == 0 && txmp.config.TTLDuration {
    return
}
```
then we can skip looping through all the elements. I think the default is actually that these TTLs are switched off but maybe double check.

File: internal/mempool/v1/mempool.go [hunk 12]
```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
...
 	}
 
+	txmp.logger.Debug(
+		"existing transaction no longer valid; failed re-CheckTx callback",
+		"priority", wtx.Priority(),
+		"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
+		"err", err,
+		"code", checkTxRes.CheckTx.Code,
+	)
+	txmp.removeTxByElement(elt)
```
Comment (cmwaters) [inline internal/mempool/v1/mempool.go:674]: Do we consider this transaction evicted? Should we record this in our metrics?

File: internal/mempool/v1/mempool.go [hunk 13]
```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
...
+		// N.B. Grab the next element first, since if we remove cur its successor
+		// will be invalidated.
+		next := cur.Next()
+
+		w := cur.Value.(*WrappedTx)
+		if txmp.config.TTLNumBlocks > 0 && (blockHeight-w.height) > txmp.config.TTLNumBlocks {
+			txmp.removeTxByElement(cur)
+		} else if txmp.config.TTLDuration > 0 && now.Sub(w.timestamp) > txmp.config.TTLDuration {
+			txmp.removeTxByElement(cur)
 		}
```
Comment (cmwaters) [inline internal/mempool/v1/mempool.go:761]: Do we want to mark these transactions as evicted in our metrics?

File: internal/mempool/v1/mempool.go (hunk 8 above)
Comment (cmwaters) [reply internal/mempool/v1/mempool.go:663]: Also what about gas wanted. Can that be updated in `ReCheckTx`?

**Review by cmwaters — COMMENTED** [review]
Nice work on this so far 🎉 . Left a few comments.

File: internal/mempool/v1/mempool.go (hunk 2 above)
Comment (creachadair) [reply internal/mempool/v1/mempool.go:512]: Probably not. I just copied the log. Do you think it should be debug or shall I just remove it entirely?

File: internal/mempool/v1/mempool.go (hunk 7 above)
Comment (creachadair) [reply internal/mempool/v1/mempool.go:641]: I'll add a comment. It's detecting that we are finished with a recheck cycle. The extra check is because we might have been flushed during the recheck, so we will still receive the callbacks (so we don't stall the client) but we don't process the tx (which is, perforce, gone)

File: internal/mempool/v1/mempool.go (hunk 8 above)
Comment (creachadair) [reply internal/mempool/v1/mempool.go:663]: Both good questions. I don't think there's any reason we couldn't change those values too.

File: internal/mempool/v1/mempool.go (hunk 9 above)
Comment (creachadair) [reply internal/mempool/v1/mempool.go:673]: I don't know. If you have suggestions I'm glad to add them; I mainly just kept the logs as they were.

File: internal/mempool/v1/mempool.go (hunk 10 above)
Comment (creachadair) [reply internal/mempool/v1/mempool.go:749]: If it becomes an issue I'll change it, this is all internal anyway.

File: internal/mempool/v1/mempool.go (hunk 11 above)
Comment (creachadair) [reply internal/mempool/v1/mempool.go:748]: That would make the code more legible anyway, so it's worth doing even though the savings may be negligible.

File: internal/mempool/v1/mempool.go (hunk 3 above)
Comment (creachadair) [reply internal/mempool/v1/mempool.go:555]: It should.

File: internal/mempool/v1/mempool.go (hunk 2 above)
Comment (cmwaters) [reply internal/mempool/v1/mempool.go:512]: Yeah I think debug makes sense

File: internal/mempool/v1/mempool.go (hunk 8 above)
Comment (cmwaters) [reply internal/mempool/v1/mempool.go:670]: Yeah I think for completeness we should update them although I don't think the SDK actually does that

File: internal/mempool/v1/mempool.go (hunk 13 above)
Comment (creachadair) [reply internal/mempool/v1/mempool.go:761]: That's a good question. I've added that for now, but this seems like a case where we might want a separate metric for these (ExpiredTransactions, say).

File: internal/mempool/v1/mempool.go (hunk 8 above)
Comment (creachadair) [reply internal/mempool/v1/mempool.go:672]: I've left this as-is for now, as the original priority mempool implementation only updated priority on recheck. But if you think we should update the others as well, that's definitely something we can do.

File: internal/mempool/v1/mempool.go (hunk 12 above)
Comment (cmwaters) [reply internal/mempool/v1/mempool.go:683]: We can have the metric semantics work one of two ways:
1. Rejected is any event that prevents a transaction from entering the mempool either because of the app or because of Tendermint and Evicted is any action after a transaction enters the mempool and causes it to be forced out.
2. Rejected is any tx that is invalid (from the application's perspective), Evicted is for those that are removed after entering and we probably need another metric to capture transactions that never made it for non-validity based reasons.

For (1) we should call this Evicted for (2) we should call is Rejected. Either is fine so long as it's documented and clear :)

Comment (creachadair) [reply internal/mempool/v1/mempool.go:683]: Unfortunately, the doc comments in the metrics type don't really allow either: "Rejected" is "transactions that passed CheckTx but failed to make it into the mempool due to resource limits"; "Evicted" is "valid transactions that passed CheckTx and existed in the mempool but were later evicted to make room for higher priority valid transactions that passed CheckTx".

We could go either way, but we have to violate the definitions in any case.  I think extending the definition of Rejected is probably the lest-invasive change here.

Comment (creachadair) [reply internal/mempool/v1/mempool.go:683]: Er, sorry, I said that backward, I mean "Evicted" is probably the least invasive.

Comment (cmwaters) [reply internal/mempool/v1/mempool.go:683]: So there are no metrics for transactions that failed CheckTx?

**Comment by creachadair** [comment]
> I think one of these test failures is real. I'll take a closer look in the morning.

Ok, this should be fixed now.

File: internal/mempool/v1/reactor.go [hunk 14]
```diff
@@ -329,6 +326,8 @@ func (r *Reactor) broadcastTxRoutine(peerID types.NodeID, closer *tmsync.Closer)
 		}
 	}()
 
+	peerMempoolID := r.ids.GetForPeer(peerID)
+	var nextGossipTx *clist.CElement
 	for {
```
Comment (tychoish) [inline internal/mempool/v1/reactor.go:331]: I don't see the condition where this loop ever breaks in the case that we see another TX

Comment (creachadair) [reply internal/mempool/v1/reactor.go:331]: It doesn't break in that case—the `nextGossipTx` is just a state variable reused during the lifecycle of the loop. All I did here was move its declaration closer to the use.

File: internal/mempool/v1/mempool.go (hunk 12 above)
Comment (creachadair) [reply internal/mempool/v1/mempool.go:683]: There appears to be a FailedTxs metric. The original code wasn't setting it, but I will add that here since it makes sense to use what we already have.

The only case I think we don't distinguish right now is "failed on recheck".

Comment (creachadair) [reply internal/mempool/v1/mempool.go:683]: (I have swapped it to that)

Comment (cmwaters) [reply internal/mempool/v1/mempool.go:683]: Yeah that makes sense to me as well

Comment (cmwaters) [reply internal/mempool/v1/mempool.go:683]: So we have rejected, failed and evicted.

Comment (creachadair) [reply internal/mempool/v1/mempool.go:683]: Correct. As of now:

- Rejected: Passed initial CheckTx, rejected due to capacity.
- Evicted: Passed initial CheckTx and added to mempool, later ejected either by pruning (for TTL) or because a higher-priority transaction took precedence.
- Failed: Failed either initial or recheck CheckTx.

**Review by cmwaters — APPROVED** [review]
👍

File: internal/mempool/v1/mempool.go [hunk 15]
```diff
@@ -450,422 +416,373 @@ func (txmp *TxMempool) Update(
...
-	wtx.sender = sender
-	wtx.peers = map[uint16]struct{}{
-		txInfo.SenderID: {},
-	}
+	wtx.SetGasWanted(checkTxRes.CheckTx.GasWanted)
+	wtx.SetPriority(priority)
+	wtx.SetSender(sender)
 
 	txmp.metrics.TxSizeBytes.Observe(float64(wtx.Size()))
 	txmp.metrics.Size.Set(float64(txmp.Size()))
```
[the remark this answers is not in the document]
Comment (creachadair) [reply internal/mempool/v1/mempool.go:602]: Yes, good catch, thanks.

**Review by williambanfield — COMMENTED** [review]
Looks good. There are a few edge cases on adding a new transaction with priority / size / gas that may be worth ensuring there's a test for.

File: internal/mempool/v1/mempool.go [hunk 16]
```diff
@@ -204,240 +145,265 @@ func (txmp *TxMempool) EnableTxsAvailable() {
...
-		gas := totalGas + wtx.gasWanted
-		if maxGas > -1 && gas > maxGas {
-			return txs[:len(txs)-1]
+// ReapMaxBytesMaxGas returns a slice of valid transactions that fit within the
+// size and gas constraints. The results are ordered by nonincreasing priority,
+// with ties broken by increasing order of arrival.  Reaping transactions does
+// not remove them from the mempool.
+//
+// If maxBytes < 0, no limit is set on the total size in bytes.
+// If maxGas < 0, no limit is set on the total gas cost.
```
[the remark this answers is not in the document]
Comment (creachadair) [reply internal/mempool/v1/mempool.go:329]: As far as I can tell that is the behaviour of the existing v0 code (though it's possible I misread it).

File: internal/mempool/v1/tx.go [hunk 17]
```diff
@@ -1,281 +1,87 @@
...
-func (txs *TxStore) IsTxRemoved(hash types.TxKey) bool {
-	txs.mtx.RLock()
-	defer txs.mtx.RUnlock()
-
-	wtx, ok := txs.hashTxs[hash]
-	if ok {
-		return wtx.removed
+	tx        types.Tx    // the original transaction data
+	hash      types.TxKey // the transaction hash
+	height    int64       // height at which this transaction was checked (for expiry)
```
[the remark this answers is not in the document]
Comment (creachadair) [reply internal/mempool/v1/tx.go:15]: Sorry, I'm not sure what you mean.

**Comment by creachadair** [comment]
> Looks good. There are a few edge cases on adding a new transaction with priority / size / gas that may be worth ensuring there's a test for.

Yes, I plan to add tests for all those corners. Nothing so far is really exercising those conditions at all.

File: internal/mempool/v1/mempool.go [hunk 18]
```diff
@@ -450,422 +416,372 @@ func (txmp *TxMempool) Update(
...
-		)
-		if len(evictTxs) == 0 {
-			// No room for the new incoming transaction so we just remove it from
-			// the cache.
+		var victims []*clist.CElement // eligible transactions for eviction
+		var victimBytes int64         // total size of victims
+		for cur := txmp.txs.Front(); cur != nil; cur = cur.Next() {
+			cw := cur.Value.(*WrappedTx)
+			if cw.priority < priority {
+				victims = append(victims, cur)
```
[the remark this answers is not in the document]
Comment (creachadair) [reply internal/mempool/v1/mempool.go:537]: Yes.

File: internal/mempool/v1/mempool.go [hunk 19]
```diff
@@ -450,422 +416,372 @@ func (txmp *TxMempool) Update(
...
 	}
 
 	txmp.purgeExpiredTxs(blockHeight)
 
 	// If there any uncommitted transactions left in the mempool, we either
 	// initiate re-CheckTx per remaining transaction or notify that remaining
 	// transactions are left.
-	if txmp.Size() > 0 {
+	size := txmp.Size()
+	txmp.metrics.Size.Set(float64(size))
```
[the remark this answers is not in the document]
Comment (creachadair) [reply internal/mempool/v1/mempool.go:438]: If we do a recheck it gets updated at the end.

File: internal/mempool/v1/tx.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (creachadair) [reply internal/mempool/v1/tx.go:15]: Oh I see. You mean vs. a recheck. Yes, I'll clarify that.

**Review by williambanfield — APPROVED** [review]
